pybamm/solvers/c_solvers/idaklu/Expressions/Base/Expression.hpp

```diff
@@ -0,0 +1,69 @@
+#ifndef PYBAMM_EXPRESSION_HPP
+#define PYBAMM_EXPRESSION_HPP
+
+#include "ExpressionTypes.hpp"
+#include "../../common.hpp"
+#include "../../Options.hpp"
+#include <memory>
+#include <vector>
+
+class Expression {
+public:  // method declarations
+  /**
+   * @brief Constructor
+   */
+  Expression() = default;
+
+  /**
+   * @brief Evaluation operator
+   */
+  virtual void operator()() = 0;
+
+  /**
+   * @brief Evaluation operator given data vectors
```

Comment:
can you please expand on the descriptions of the inputs/outputs? You mention they are vectors, but then later functions indicate that the output can be matrices?

Reply:
This refers to a 'vector' of arguments (input data), where each argument can be of arbitrary size (although in reality this is limited within the solver). I have rephrased the docstrings to `@brief Evaluation operator (for use after setting input and output data references)` and `Evaluation operator (supplying data references)` as-well as clarifying `get_rows()` and `get_cols()` to e.g. `Returns row indices in COO format (where the output data represents sparse matrix elements)`. Note that the spartsity pattern is computed in Python before calling the solver and is assumed fixed.